File: C++/Array/Sorting/Insertionsort.cpp
//Insertion Sort->
//Insert an element from unsorted array to its correct position in sorted array.
//Start from Index[1] leaving Index[0] aside.
//Imagine sorting a deck of cards.
#include<iostream>
using namespace std;

int main(){
	int arr[] = {23, 12, 99, 7, 54, 65};
	int size = sizeof(arr) / sizeof(arr[0]);
	
	for(int i = 1; i < size; i++){//This loop will select an element which will be needed to sorted.
		int current = arr[i];//That element is stored here.
		int j = i - 1;
		while(arr[j] > current && j >= 0){//This loop will compare 'current' element to all the element on its left and insert
						  //it to its right place.
			arr[j+1] = arr[j];
			j--;
		}
		arr[j+1] = current;
	}
	for(int i = 0; i < size; i++){
		cout<<arr[i]<<" ";
	}
	return 0;
}

File: Java/ARRAY/Sorting/insertion sort.java
public class Main{
    public static void insertionSort(int arr[]){
        for(int i = 1; i < arr.length; i++){
            int curr = arr[i];
            int prev = i - 1;
            
            // Here, we are checking if the current element is shorter than the element on the left side of it. If so shift those elements to the right side to create space for the new element.
            while(prev >= 0 && arr[prev] > curr){
                arr[prev + 1] = arr[prev];
                prev--;
            }
            
            // Here, we are placing the current element to its right position
            arr[prev+1] = curr;
        }
    }
    
    public static void printArray(int arr[]){
        for(int i = 0; i < arr.length; i++){
            System.out.print(arr[i] + " ");
        }
    }
    public static void main (String[] args) {
        int arr[] = {5, 4, 1,3,2};
        insertionSort(arr);
        printArray(arr);
    }
}